compiler/dyno/lib/resolution/resolution-queries.cpp

```diff
@@ -619,8 +619,8 @@ const Type* initialTypeForTypeDecl(Context* context, ID declId) {
 const ResolvedFields& resolveFieldDecl(Context* context,
                                        const CompositeType* ct,
                                        ID fieldId,
-                                       bool useGenericFormalDefaults) {
-  QUERY_BEGIN(resolveFieldDecl, context, ct, fieldId, useGenericFormalDefaults);
+                                       DefaultsPolicy defaultsPolicy) {
+  QUERY_BEGIN(resolveFieldDecl, context, ct, fieldId, defaultsPolicy);
```

Comment:
Since this function resolves one field (or one field multi-decl), I was expecting some code near here to set defaultsPolicy=USE_DEFAULT_OTHER_FIELDS and I'm not understanding how it is working without that.

Reply:
With the example from the OP (duplicated to improve locality 😄 ) 

```Chapel
proc id(type t) type return t;

record r {
  type typeWithDefault = int;
  var nonGenericField : id(typeWithDefault);
}
```

If we resolved each field (individually) of `r` using `USE_DEFAULT_OTHER_FIELDS`, we'd end up with `typeWithDefault : type AnyType`, but `nonGenericField : int` (because when resolving `nonGenericField` with `USE_DEFAULT_OTHER_FIELDS`, `typeWithDefault` is another field, so we'd use its defaults). If we wanted to get the "least generic" version of `r`, that result is incorrect; the "least generic" version of `r` has fields with `type AnyType` and `Unknown`, or something along those lines (certainly `int` shouldn't occur). So we don't want to override `defaultsPolicy` here.

More generally, the three policies as far as I'm concerned are:
* `IGNORE_DEFAULTS` to create the most generic version of the type (no defaults at all)
* `USE_DEFAULT_OTHER_FIELDS` to determine which fields are generic in their own right (as described in the OP)
* `USE_DEFAULTS` to instantiate a type with its defaults

The `resolveFieldDecl` is required in all of these three cases. The `defaultsPolicy` is set outside of this query depending on the use case. In `getTypeWithDefaults`, we use `OTHER_FIELDS` and `USE` to identify fields for which only _their own_ default makes a difference (and thus a substitution is required). In `getFieldsGenericity`, we also use `OTHER_FIELDS`, as it helps identify fields that are generic "in their own right". If we can't instantiate the type (it's not generic with defaults) we use `IGNORE_DEFAULTS`.